engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java

```diff
@@ -45,38 +47,46 @@ public StateMainMenu(String showMessageOnLoad) {
     @Override
     public void init(GameEngine gameEngine) {
         context = gameEngine.createChildContext();
-        initEntityAndComponentManagers(false);
+        headless = context.get(DisplayDevice.class).isHeadless();
+
+        initEntityAndComponentManagers(headless);
 
         createLocalPlayer(context);
 
-        // TODO: REMOVE this and handle refreshing of core game state at the engine level - see Issue #1127
-        new RegisterInputSystem(context).step();
+        if (!headless) {
+            // TODO: REMOVE this and handle refreshing of core game state at the engine level - see Issue #1127
+            new RegisterInputSystem(context).step();
 
-        nuiManager = context.get(NUIManager.class);
-        eventSystem.registerEventHandler(nuiManager);
-        NUIEditorSystem nuiEditorSystem = new NUIEditorSystem();
-        context.put(NUIEditorSystem.class, nuiEditorSystem);
-        componentSystemManager.register(nuiEditorSystem, "engine:NUIEditorSystem");
+            nuiManager = context.get(NUIManager.class);
+            eventSystem.registerEventHandler(nuiManager);
+            NUIEditorSystem nuiEditorSystem = new NUIEditorSystem();
+            context.put(NUIEditorSystem.class, nuiEditorSystem);
+            componentSystemManager.register(nuiEditorSystem, "engine:NUIEditorSystem");
 
-        NUISkinEditorSystem nuiSkinEditorSystem = new NUISkinEditorSystem();
-        context.put(NUISkinEditorSystem.class, nuiSkinEditorSystem);
-        componentSystemManager.register(nuiSkinEditorSystem, "engine:NUISkinEditorSystem");
+            NUISkinEditorSystem nuiSkinEditorSystem = new NUISkinEditorSystem();
+            context.put(NUISkinEditorSystem.class, nuiSkinEditorSystem);
+            componentSystemManager.register(nuiSkinEditorSystem, "engine:NUISkinEditorSystem");
 
-        inputSystem = context.get(InputSystem.class);
+            inputSystem = context.get(InputSystem.class);
+        }
 
         componentSystemManager.initialise();
-
         console = context.get(Console.class);
         storageServiceWorker = context.get(StorageServiceWorker.class);
-
         playBackgroundMusic();
 
-        //guiManager.openWindow("main");
-        context.get(NUIManager.class).pushScreen("engine:mainMenuScreen");
+        if (!headless) {
+            //guiManager.openWindow("main");
+            context.get(NUIManager.class).pushScreen("engine:mainMenuScreen");
+        }
         if (!messageOnLoad.isEmpty()) {
             TranslationSystem translationSystem = context.get(TranslationSystem.class);
-            MessagePopup popup = nuiManager.pushScreen(MessagePopup.ASSET_URI, MessagePopup.class);
-            popup.setMessage("Error", translationSystem.translate(messageOnLoad));
+            if (headless) {
+                throw new RuntimeException("Game receive error and fall to main menu. See logs before");
```

Comment:
Let's add messageOnLoad to this exception. Yes, something else _probably_ logged something when it set this messageOnLoad field, but I often have trouble actually finding those messages.